tests/benchmark/test_engine.py

```diff
@@ -0,0 +1,191 @@
+from aiida.plugins.factories import CalculationFactory
+
+ArithmeticAddCalculation = CalculationFactory('arithmetic.add')
+
+
+class WorkchainLoop(WorkChain):
+    """A basic Workchain to run a looped step n times."""
+
+    @classmethod
+    def define(cls, spec):
+        super().define(spec)
+        spec.input('iterations', required=True)
+        spec.input('code', required=False)
+        spec.outline(cls.init_loop, while_(cls.terminate_loop)(cls.run_task))
+        # spec.output('result')
+
+    def init_loop(self):
+        self.ctx.iter = self.inputs.iterations.value
+        self.ctx.counter = 0
+
+    def terminate_loop(self):
+        if self.ctx.counter >= self.ctx.iter:
+            return False
+        self.ctx.counter += 1
+        return True
+
+    def run_task(self):
+        pass
+
+
+class WorkchainLoopWcSerial(WorkchainLoop):
+    """A WorkChain that submits another WorkChain n times in different steps."""
+
+    def run_task(self):
+        future = self.submit(WorkchainLoop, iterations=Int(1))
+        return ToContext(**{'wkchain' + str(self.ctx.counter): future})
+
+
+class WorkchainLoopWcThreaded(WorkchainLoop):
+    """A WorkChain that submits another WorkChain n times in the same step."""
+
+    def init_loop(self):
+        super().init_loop()
+        self.ctx.iter = 1
+
+    def run_task(self):
+
+        context = {
+            'wkchain' + str(i): self.submit(WorkchainLoop, iterations=Int(1))
+            for i in range(self.inputs.iterations.value)
+        }
+        return ToContext(**context)
+
+
+class WorkchainLoopCalcSerial(WorkchainLoop):
+    """A WorkChain that submits a CalcJob n times in different steps."""
+
+    def run_task(self):
+        inputs = {
+            'x': Int(1),
+            'y': Int(2),
+            'code': self.inputs.code,
+        }
+        future = self.submit(ArithmeticAddCalculation, **inputs)
+        return ToContext(addition=future)
+
+
+class WorkchainLoopCalcThreaded(WorkchainLoop):
+    """A WorkChain that submits a CalcJob n times in the same step."""
+
+    def init_loop(self):
+        super().init_loop()
+        self.ctx.iter = 1
+
+    def run_task(self):
+        futures = {}
+        for i in range(self.inputs.iterations.value):
+            inputs = {
+                'x': Int(1),
+                'y': Int(2),
+                'code': self.inputs.code,
+            }
+            futures['addition' + str(i)] = self.submit(ArithmeticAddCalculation, **inputs)
+        return ToContext(**futures)
+
+
+WORKCHAINS = {
+    'basic-loop': (WorkchainLoop, 4, 0),
+    'serial-wc-loop': (WorkchainLoopWcSerial, 4, 4),
+    'threaded-wc-loop': (WorkchainLoopWcThreaded, 4, 4),
+    'serial-calcjob-loop': (WorkchainLoopCalcSerial, 4, 4),
+    'threaded-calcjob-loop': (WorkchainLoopCalcThreaded, 4, 4),
+}
+
+
+@pytest.mark.parametrize('workchain,iterations,outgoing', WORKCHAINS.values(), ids=WORKCHAINS.keys())
+@pytest.mark.usefixtures('clear_database_before_test')
```

Comment:
what is the advantage of adding it this way rather than making it an argument of the test function like all other fixtures?

Note: one might be able to turn on this fixture for benchmark tests automatically by placing a `conftest.py` with content
```python
import pytest

@pytest.fixture(scope='function', autouse=True)
def clear(clear_database_before_test):
    pass
```
but I'm not sure and perhaps it's more confusing than helpful.

Reply:
Main argument is that this fixture is not actually supposed to be called by the test function itself. It should just be injected. You could add it as an argument, but then linters, such as pylint, will think you made a mistake and issue a warning. This approach makes it clearer that it is a non-callable fixture

Reply:
Would be good to decide whether we want to adopt this approach going forward - both to know what to follow in new tests and because it means that it is no longer enough to look at the test function arguments to understand which fixtures were called.

Reply:
I have actually already been using this in all the pytests I have added, so I am definitely in favor. I don't see the problem in having the fixtures in a decorator if they don't get explicitly called in the test body.

Reply:
Yep, it's a standard pytest practice, so I don't see any issue with using it.
If you want to understand what fixtures are being called, you shouldn't look at the function anyway, you call `pytest --setup-plan` 😉